File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/options/appearance/Language.java
package wangdaye.com.geometricweather.common.basic.models.options.appearance;

import android.content.Context;
import android.content.res.Resources;
import android.os.Build;
import android.text.TextUtils;

import androidx.annotation.Nullable;

import java.util.Locale;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.basic.models.options._utils.Utils;

public enum Language {

    FOLLOW_SYSTEM(
            "follow_system",
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
                    ? Resources.getSystem().getConfiguration().getLocales().get(0)
                    : Resources.getSystem().getConfiguration().locale
    ),
    CHINESE("chinese", new Locale("zh", "CN")),
    UNSIMPLIFIED_CHINESE("unsimplified_chinese", new Locale("zh", "TW")),
    ENGLISH_US("english_america", new Locale("en", "US")),
    ENGLISH_UK("english_britain", new Locale("en", "GB")),
    ENGLISH_AU("english_australia", new Locale("en", "AU")),
    TURKISH("turkish", new Locale("tr")),
    FRENCH("french", new Locale("fr")),
    RUSSIAN("russian", new Locale("ru")),
    GERMAN("german", new Locale("de")),
    SERBIAN("serbian", new Locale("sr")),
    SPANISH("spanish", new Locale("es")),
    ITALIAN("italian", new Locale("it")),
    DUTCH("dutch", new Locale("nl")),
    HUNGARIAN("hungarian", new Locale("hu")),
    PORTUGUESE("portuguese", new Locale("pt")),
    PORTUGUESE_BR("portuguese_brazilian", new Locale("pt", "BR")),
    SLOVENIAN("slovenian", new Locale("sl", "SI")),
    ARABIC("arabic", new Locale("ar")),
    CZECH("czech", new Locale("cs")),
    POLISH("polish", new Locale("pl")),
    KOREAN("korean", new Locale("ko")),
    GREEK("greek", new Locale("el")),
    JAPANESE("japanese", new Locale("ja")),
    ROMANIAN("romanian", new Locale("ro"));

    private final String languageId;
    private final Locale locale;

    Language(String languageId, Locale locale) {
        this.languageId = languageId;
        this.locale = locale;
    }

    public String getCode() {
        Locale locale = getLocale();
        String language = locale.getLanguage();
        String country = locale.getCountry();

        if (!TextUtils.isEmpty(country)
                && (country.toLowerCase().equals("tw") || country.toLowerCase().equals("hk"))) {
            return language.toLowerCase() + "-" + country.toLowerCase();
        } else {
            return language.toLowerCase();
        }
    }

    public Locale getLocale() {
        return locale;
    }

    @Nullable
    public String getLanguageName(Context context) {
        return Utils.getNameByValue(
                context.getResources(),
                languageId,
                R.array.languages,
                R.array.language_values
        );
    }

    public boolean isChinese() {
        return getCode().startsWith("zh");
    }

    public static Language getInstance(String value) {
        switch (value) {
            case "chinese":
                return CHINESE;

            case "unsimplified_chinese":
                return UNSIMPLIFIED_CHINESE;

            case "english_america":
                return ENGLISH_US;

            case "english_britain":
                return ENGLISH_UK;

            case "english_australia":
                return ENGLISH_AU;

            case "turkish":
                return TURKISH;

            case "french":
                return FRENCH;

            case "russian":
                return RUSSIAN;

            case "german":
                return GERMAN;

            case "serbian":
                return SERBIAN;

            case "spanish":
                return SPANISH;

            case "italian":
                return ITALIAN;

            case "dutch":
                return DUTCH;

            case "hungarian":
                return HUNGARIAN;

            case "portuguese":
                return PORTUGUESE;

            case "portuguese_brazilian":
                return PORTUGUESE_BR;

            case "slovenian":
                return SLOVENIAN;

            case "arabic":
                return ARABIC;

            case "czech":
                return CZECH;

            case "polish":
                return POLISH;

            case "korean":
                return KOREAN;

            case "greek":
                return GREEK;

            case "japanese":
                return JAPANESE;

            case "romanian":
                return ROMANIAN;

            default:
                return FOLLOW_SYSTEM;
        }
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/options/appearance/Language.kt
package wangdaye.com.geometricweather.common.basic.models.options.appearance

import android.content.Context
import android.content.res.Resources
import android.os.Build
import android.text.TextUtils
import wangdaye.com.geometricweather.R
import wangdaye.com.geometricweather.common.basic.models.options._basic.BaseEnum
import wangdaye.com.geometricweather.common.basic.models.options._basic.Utils
import java.util.*

enum class Language(
    override val id: String,
    val locale: Locale
): BaseEnum {

    FOLLOW_SYSTEM(
        "follow_system",
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            Resources.getSystem().configuration.locales[0]
        } else {
            Resources.getSystem().configuration.locale
        }
    ),
    CHINESE("chinese", Locale("zh", "CN")),
    UNSIMPLIFIED_CHINESE("unsimplified_chinese", Locale("zh", "TW")),
    ENGLISH_US("english_america", Locale("en", "US")),
    ENGLISH_UK("english_britain", Locale("en", "GB")),
    ENGLISH_AU("english_australia", Locale("en", "AU")),
    TURKISH("turkish", Locale("tr")),
    FRENCH("french", Locale("fr")),
    RUSSIAN("russian", Locale("ru")),
    GERMAN("german", Locale("de")),
    SERBIAN("serbian", Locale("sr")),
    SPANISH("spanish", Locale("es")),
    ITALIAN("italian", Locale("it")),
    DUTCH("dutch", Locale("nl")),
    HUNGARIAN("hungarian", Locale("hu")),
    PORTUGUESE("portuguese", Locale("pt")),
    PORTUGUESE_BR("portuguese_brazilian", Locale("pt", "BR")),
    SLOVENIAN("slovenian", Locale("sl", "SI")),
    ARABIC("arabic", Locale("ar")),
    CZECH("czech", Locale("cs")),
    POLISH("polish", Locale("pl")),
    KOREAN("korean", Locale("ko")),
    GREEK("greek", Locale("el")),
    JAPANESE("japanese", Locale("ja")),
    ROMANIAN("romanian", Locale("ro"));

    val code: String
        get() {
            val locale = locale
            val language = locale.language
            val country = locale.country
            return if (!TextUtils.isEmpty(country)
                && (country.lowercase() == "tw" || country.lowercase() == "hk")
            ) {
                language.lowercase() + "-" + country.lowercase()
            } else {
                language.lowercase()
            }
        }

    val isChinese: Boolean
        get() = code.startsWith("zh")

    companion object {

        fun getInstance(
            value: String
        ) = when (value) {
            "chinese" -> CHINESE
            "unsimplified_chinese" -> UNSIMPLIFIED_CHINESE
            "english_america" -> ENGLISH_US
            "english_britain" -> ENGLISH_UK
            "english_australia" -> ENGLISH_AU
            "turkish" -> TURKISH
            "french" -> FRENCH
            "russian" -> RUSSIAN
            "german" -> GERMAN
            "serbian" -> SERBIAN
            "spanish" -> SPANISH
            "italian" -> ITALIAN
            "dutch" -> DUTCH
            "hungarian" -> HUNGARIAN
            "portuguese" -> PORTUGUESE
            "portuguese_brazilian" -> PORTUGUESE_BR
            "slovenian" -> SLOVENIAN
            "arabic" -> ARABIC
            "czech" -> CZECH
            "polish" -> POLISH
            "korean" -> KOREAN
            "greek" -> GREEK
            "japanese" -> JAPANESE
            "romanian" -> ROMANIAN
            else -> FOLLOW_SYSTEM
        }
    }

    override val valueArrayId = R.array.language_values
    override val nameArrayId = R.array.languages

    override fun getName(context: Context) = Utils.getName(context, this)
}